src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+                        .Except(instantiatedTypes)
+                        .Where(type => !HasInstantiatedNestedType(type, instantiatedTypes));
+
+                    foreach (INamedTypeSymbol type in uninstantiatedInternalTypes)
+                    {
+                        context.ReportDiagnostic(type.CreateDiagnostic(Rule, type.FormatMemberName()));
+                    }
+                });
+            });
+        }
+
+        private bool AssemblyExposesInternals(Compilation compilation)
+        {
+            ISymbol assemblySymbol = compilation.Assembly;
+            ImmutableArray<AttributeData> attributes = assemblySymbol.GetAttributes();
+            return attributes.Any(
+                attr => attr.AttributeClass.Name.Equals("InternalsVisibleToAttribute", StringComparison.Ordinal));
+        }
+
+        private bool HasInstantiatedNestedType(INamedTypeSymbol type, List<INamedTypeSymbol> instantiatedTypes)
+        {
+            // We don't care whether a private nested type is instantiated, because if it
+            // is, it can only have happened within the type itself.
+            IEnumerable<INamedTypeSymbol> nestedTypes = type.GetTypeMembers()
+                .Where(member => member.DeclaredAccessibility != Accessibility.Private);
+
+            foreach (var nestedType in nestedTypes)
+            {
+                if (instantiatedTypes.Contains(nestedType))
+                {
+                    return true;
+                }
+
+                if (HasInstantiatedNestedType(nestedType, instantiatedTypes))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public static bool IsOkToBeUnused(INamedTypeSymbol type, Compilation compilation)
+        {
+            if (type.TypeKind != TypeKind.Class || type.IsAbstract)
+            {
+                return true;
+            }
+
+            // Attributes are not instantiated in IL but are created by reflection.
+            INamedTypeSymbol attributeSymbol = compilation.GetTypeByMetadataName("System.Attribute");
+            if (type.Inherits(attributeSymbol))
+            {
+                return true;
+            }
+
+            // The type containing the assembly's entry point is OK.
+            if (ContainsEntryPoint(type, compilation))
+            {
+                return true;
+            }
+
+            // MEF exported classes are instantiated by MEF, by reflection.
+            if (IsMefExported(type, compilation))
+            {
+                return true;
+            }
+
+            // Types implementing the (deprecated) IConfigurationSectionHandler interface
+            // are OK because they are instantiated by the configuration system.
+            INamedTypeSymbol iConfigurationSectionHandlerSymbol = compilation.GetTypeByMetadataName("System.Configuration.IConfigurationSectionHandler");
```

Comment:
Same here - in fact GetTypeByMetadataName should only be used in the compilation start action.